### DISCUSSION TITLE
formRef getModel documentation

### DISCUSSION THREAD

**Discussion by simplecommerce** [vazco/uniforms#1116] [Ideas]
Hi,

I noticed that the `getModel` method that is avaiable via `formRef` is not in your documentation.
I just thought maybe it should be, in my case I found it very useful as a method in order to get the updated model from outside the form, especially with the `mode` argument which allows me to get the transformed model.

**Comment by radekmie** [comment]
Hi @simplecommerce. While I agree that it's much better than simply accessing the props or state of the form, I'm still unsure whether or not we want to make it a part of the public API. Could you share your exact use case?

**Comment by simplecommerce** [comment]
@radekmie No problem, here are my current use cases.

In one case, I need to compare the saved value from my form in my database against the calculated value in my form.
Flow:
1. An amount is received from my database. (e.g.: Subtotal)
2. In my form's model, I do some specific calculations and the result is stored in a field. (Total)
3. This field is saved in my database.
4. When I go back to the form, the saved value is reloaded in the form. (Total)
5. If the amount that is received from my database has changed (Subtotal), then the calculated value (Total) will differ than what was stored initially.

So I have a useEffect on mount, that validates the saved Total vs the new calculated Total by using the `getModel("submit")`.

In another case, I have some logic in the `onChange` method of the form.
I check for a specific field that is being changed, and then I also need to compare other values from the model itself, before applying my logic. So I use the `getModel()` to get that information.

For this case, I could've put the logic in each component, but since the logic was similar but just the field names changed, I didn't want to repeat my code/logic in each component and centralized it in the form's onChange method.

In another case, I have a text field that has a calculate button. When I click that button, I need to pull information from the model in order to apply the calculations needed. The reason I am using `getModel` here, is that my text field is already a custom component that uses `useField` internally and the calculate button is an adornment option that I am using (Material-UI) in order to add additional feature to it for a specific case. 

Sorry if I am not too clear with my examples, but in my cases I just found it easier to implement when using your `getModel` method. (in my mind)

**Reply by radekmie** [reply]
I think these are completely valid use cases for this function. We'll discuss how to approach it.

**Reply by wadamek65** [reply]
Since we could not come up with an idea for a cleaner API for this, I think it would be best to document this with a note mentioning that this pattern should be avoided whenever possible.

**Reply by simplecommerce** [reply]
@wadamek65 what would be your suggestion in getting the internal model values of the form from the outside?

**Reply by wadamek65** [reply]
@simplecommerce As I've mentioned, we couldn't really come up with a better idea for a solution for this problem so what you did is probably optimal. But this (imperative) pattern IMO isn't very idiomatic to React and should be avoided if there is a better one available.

**Reply by radekmie** [reply]
Just to confirm what @wadamek65 said - there's no other API for that and we don't want to document the one that is there now. If we'll come up with one we think is better, it'll definitely get documented!

(If you have an idea for it, do let us know!)
